Modify MPAS-Ocean Registry files to have CF compliant units

darincomeau opened the pull request:
This modifies MPAS-Ocean Registry files to have CF compliant units, using `cfchecker` to find errors. 

[CF Standard Units](https://cfconventions.org/Data/cf-standard-names/current/build/cf-standard-name-table.html)

[WIP]

darincomeau commented:
Some remaining errors from using `cfchecker` on `timeSeriesStatsDaily` test case:

1. units="NA" or "nondimensional". I believe some of these should be "1", while others should have no units attribute, so this probably needs to be looked at individually.
2. units with PSU.
3. `xtime_startDaily`, `xtime_endDaily`, and `timeDaily_counter` have units="unitless", but these do not appear in mpas-ocean Registry files.

rljacob commented:
In EAM and ELM, we gave dimensionless variables a unit of "1" which is CF compliant.  That seemed better than no "units" attribute at all.  Not sure if we want an E3SM-wide standard for that or not.

xylar commented:
> In EAM and ELM, we gave dimensionless variables a unit of "1" which is CF compliant. That seemed better than no "units" attribute at all. Not sure if we want an E3SM-wide standard for that or not.

As @czender explained it to us, there are variables (like slope or other dimensionless ratios) where a unit of "1" makes sense.  But there are lots of variables where no unit (even "1") would make sense, like an integer or logical field.  So I would think places with `units="NA"` should not have a `units` attribute and ones with `nondimentional` or `unitless` might go either way.  I agree with @darincomeau that they really need to be dealt with individually.

xylar commented:
> xtime_startDaily, xtime_endDaily, and timeDaily_counter have units="unitless", but these do not appear in mpas-ocean Registry files.

@darincomeau, I can deal with these as part of my work on the `Time` coordinate if you like.

darincomeau commented:
Thanks for the comments @xylar . There are only a handful of `units="NA" or "nondimensional"`, but I was indiscriminate about wiping out all `units="unitless"` in this branch, so those may need more careful consideration.

For the time coordinate units, I haven't been able to find it from a cursory grep of framework yet. So either you or I could do it, but either way seems like it would be a separate branch from this one with strictly ocn Registry changes.

darincomeau commented:
Thanks @xylar for the pointer, that was indeed the place. Now down to just 5 NA/nondimensional errors...

darincomeau commented:
Discussion items from this morning's hackathon meeting:

1. `PSU` has been removed from all units attributes in the files changed so far, replaced to `1.e-3`. Descriptions have not been updated, with the exception of the salinity tracer, where "grams salt per kilogram seawater" was added. Question for the ocean developers - is it appropriate/desired to add PSU to the description attributes?
2.  Regarding `unit="unitless"` removal:

- many of the instances removed so far from Registries are `nml` records. They have no impact on CF compliance, but offer no useful information; I can add them back in if there's a preference to, but to me it seems cleaner to keep them out.
- there's at least one example I saw where `units="1"` would be appropriate (something like `maxCFL`). I'll go through more carefully and try to compile a list of where I've added this 1 unit designation as part of the review.

darincomeau commented:
Question: Should dimensions that are clearly integers (e.g. `nCells`) have the `units="1"` attribute? Or no `units` attribute?

czender commented:
@darincomeau Attributes can only be attached to netCDF variables, not dimensions, so I do not understand the context of your question. If there is somewhere a (coordinate) variable named `nCells`, with underlying dimension `nCells`?

xylar commented:
A agree with @czender.  If the question refers to an index array instead (e.g. `cellsOnCell`), it should definitely not have units.  I can't think of any integer or logical array that would usefully have units.

darincomeau commented:
@czender @xylar thanks for the comments. It seems like these shouldn't have been introduced at all, and this branch is clearing them. I don't see how to link file changes here, but if you look at the Files Changed in this PR, and open up `Registry.xml`, you'll see at the top lines where I've removed `units="unitless"` previously tied to dimensions.
So if I understand correctly this *is* a desired change, and these *should* be removed.

xylar commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -2,90 +2,90 @@
 
 	<dims>
-		<dim name="nCells" units="unitless"
+		<dim name="nCells"
 			 description="The number of polygons in the primary grid."
```
@darincomeau, yep, this is definitely a desired change.  To comment, you want to go to `Files changed` and use the blue plus sign to highlight one or more lines.  Then, type your comment.

xylar replied:
There isn't another way to link to the diff as far as I know.  You can link to the file itself.

xylar commented:
@darincomeau, could you rebase when you get a chance to handle the 2 conflicting files (and make sure no new non-CF units were added in the meantime)?

xylar commented:
@vanroekel, @milenaveneziani, @mark-petersen could the 3 of you have a look at these changes?  Let us know if there are questions.  The gist is that CF compliance requires that 
* variables not have a units attribute if they are unitless (i.e. no `units='unitless'`, `units='NA'`, etc.)
* variables should have `units=1` if they are dimensionless (e.g. real-valued coefficients)
* we cannot use latex symbols like `{` and `}` in the units
* we cannot use `units='PSU'` for salinity, we need to use `units=1e-3`, though we can state that the units are `grams per kilogram` in the description (used as the `long_name` attribute in output files).

@darincomeau, please edit if I missed anything.

vanroekel reviewed (approved):
These changes seem consistent with the CF compliance defined in the comment.  This is a great piece of work, lots of very careful combing through registry.

vanroekel commented:
@xylar or @darincomeau a stupid question -- does the Registry name, units, etc... populate the netcdf output metadata?  I guess that makes sense, but I had originally thought that was just for the users guide.

xylar commented:
@vanroekel, yes, many of those things end up as attributes in the NetCDF output.  Since we aren't doing a good job at keeping the user's guide updated, that is the primary thing they do at this point.

czender commented:
Sverdrup (but not Sv) is well-defined by UDUnits and thus CF:
```
zender@spectral:~$ udunits2
You have: Sverdrup
You want: m3 s-1
    1 Sverdrup = 1e+06 (m3 s-1)
    x/(m3 s-1) = 1e+06*(x/Sverdrup)
```
petawatts is also well-defined by UDUnits and CF, so I do not know what the problem with the checker is, but it is acceptable to use "petawatts" as a unit.
```
You have: petawatts
You want: watt
    1 petawatts = 1e+15 watt
    x/watt = 1e+15*(x/petawatts)

```

darincomeau commented:
Thanks @xylar , @czender . I haven't actually run any variables with units "petawatts" through the checker yet, I only tried searching for that and Sv in the CF Standard Names table and didn't see them. "Sv" appears only twice `Registry_transect_transport`, so I'll change that to Sverdrup, and leave petawatts alone (only appears 4 times in `Registry_meridional_heat_transport`.

Regarding `units=""`,  there's actually only a handful of them as well, and looks like most should be `1`, so I'll do those in a separate commit now.

darincomeau commented:
...and just found some `units="none"` and `units="None"`

xylar commented:
Thanks @czender.

darincomeau commented:
welp, i messed up the rebase

xylar commented:
Yeah, I can update `E3SM-Ocean-Discussion/E3SM/master` again but that's not necessary.  Go ahead with the E3SM PR.

mark-petersen reviewed (approved):
This looks consistent with the CF-compliant changes that we've discussed. Thanks for your work on this.

xylar commented:
@darincomeau, can we close this now that we have https://github.com/E3SM-Project/E3SM/pull/5027?  I get 2 emails every time you push something to this branch.

darincomeau commented:
Closing per https://github.com/E3SM-Project/E3SM/pull/5027
